### ISSUE TITLE
Concatenation Rule #1 is worded incorrectly

### ISSUE THREAD

**Issue by shawnisingh** [AOMediaCodec/iamf#168]
I'm fairly certain this is just a wording mistake, and not a misunderstanding of the intended semantics.

The definition says "new timeline after a Sync OBU is shifted as early as possible such that the earliest substream or parameter in the new timeline concatenates with its counterpart in the previous timeline".   This is not what was intended.  It should say something more like:  "new timeline after a Sync OBU is shifted as early as possible such that there are no overlaps in audio frames or non-redudndant parameter blocks."

With the incorrect wording, the concatenation would only look at the "earliest substream", regardless of whether any other substream would cause overlaps.  With the proposed better wording, the concatenation would look at the substream or parameter that "touches first" if we imagine holding the two timelines apart and then gradually bringing them closer together.

**Comment by cconcolato** [comment]
It is unclear to me how you can bring them closer if they already touch without creating an overlap.

**Comment by sunghee-hwang** [comment]
We also need to consider this case.
- For example, substream #2 (e.g. 100ms ~ 200ms) starts where substream #1 (e.g. 0 to 100ms) ends at.
So that, OBU sequences are Descriptor OBUs / Sync 1 / A1 / Sync 2 / A2, where A1 is audio frame for substream #1 and A2 for substream #2.

**Comment by sunghee-hwang** [comment]
In Section 4.1.12
```
relative_offset is the offset that is applied to the first audio frame or parameter block with the referenced obu_id that comes 
after this Sync OBU. It describes the position of audio and parameters in a local frame of reference. The local frame of reference
is unique for each Sync OBU.
```
In Section 6.2
```
Within a Sync OBU, only the relative information between the relative offsets is meaningful for positioning it within the global
frame of reference, where the method of positioning is described further below. This is not affected by any constant shift in the
relative offsets. As such, each Sync OBU can have any number of variants, as long as there is a constant difference between the
two variants (see the example below). This removes any restrictions on how the absolute values of the relative offsets are
selected. For example, some implementations may wish to always set the relative offset of an arbitrarily selected substream or 
parameter to 0.
```
The definition of the relative_offset is true in most case. 
But, it is not true when the current Sync OBU has only new IDs. Because there would no exact link information to concatenate between the timeline of the previous Sync OBU and the timeline of the current Sync OBU.
So, in this case the relative information should be something be able to calculate on global timeline from the timeline of the previous Sync OBU.

**Comment by sunghee-hwang** [comment]
FYI, I will make a PR for this at the next Monday (KST). If you have any comment, please let me know it before then.

**Comment by shawnisingh** [comment]
The original purpose of introducing Sync OBU was so that the relative positioning of all substreams and parameters can be understood where the Sync OBU exists, with no other context.  But the proposed changes cannot allow a playback system to join mid-stream - how will it know the end_timestamp[] without having started from the beginning?   

If we have to add such "global timeline" information also, then we might as well change Sync OBU to a "global frame of reference" instead of "local frame of reference".   But I still think there is great value to the "local frame of reference" - (1) it can save bits avoiding large timestamp numbers with leb128, and (2) it enables some interesting possibilities of splicing or manipulating IAC bitstreams with far less timestamp rewriting.   So this makes me believe the original Sync OBU approach with local frame of reference is still worth pursuing.


> We also need to consider this case.
> 
> * For example, substream 2 (e.g. 100ms ~ 200ms) starts where substream 1 (e.g. 0 to 100ms) ends at.
>   So that, OBU sequences are Descriptor OBUs / Sync 1 / A1 / Sync 2 / A2, where A1 is audio frame for substream 1 and A2 for substream 2.

I agree that concatenation rule #1 will not work when all substreams/parameters are different after the next Sync OBU.   Instead of removing local frame of reference and concatenation rule, there may be a different solution:  in these situations, concatenation rule #2 should be required.  Concat Rule #2 does not require there to be at least one substream/parameter that exists on both sides of the Sync OBU.   Global offset, can still be used to position the parameters/audio correctly in that case.  Combined with the requirement that all frames are required in timeline order, this should still be safe to avoid buffering issues.

What do you think?

**Comment by sunghee-hwang** [comment]
> The original purpose of introducing Sync OBU was so that the relative positioning of all substreams and parameters can be understood where the Sync OBU exists, with no other context. But the proposed changes cannot allow a playback system to join mid-stream - how will it know the end_timestamp[] without having started from the beginning?

The first Sync OBU after joining mid-stream is the Nth Sync OBU (N=1). So, end_timestamp[ID][0] = 0 for all ID.

**Comment by shawnisingh** [comment]
> The first Sync OBU after joining mid-stream is the Nth Sync OBU (N=1). So, end_timestamp[ID][0] = 0 for all ID.

If a playback system joins mid stream, how would it know the value of end_timestamp[ID][N-1] ?

**Comment by sunghee-hwang** [comment]
> > The first Sync OBU after joining mid-stream is the Nth Sync OBU (N=1). So, end_timestamp[ID][0] = 0 for all ID.
> 
> If a playback system joins mid stream, how would it know the value of end_timestamp[ID][N-1] ?

end_timestamp[ID][N-1] = 0.

**Comment by shawnisingh** [comment]
> end_timestamp[ID][N-1] = 0. For the first Sync OBU after join mid stream, N = 1.

Let's take two asynchronous substreams:  ID=0 with 960 ticks per frame @ 48 kHz, and ID=1 with 1024 ticks per frame @ 48 kHz.

So the timeline for first 3 frames:
ID=0 --> 0, 960, 1920
ID=1 --> 0, 1024, 2048

In absolute time after frame N=2 (after the third frames) then 
 end_timestamp[ID=0][N=2] = 2880
 end_timestamp[ID=1][N=2] = 3072.

If a playback system joins on frame N=3, then playback system will assume
  end_timestamp[ID=0][N=2] = 0
  end_timestamp[ID=1][N=2] = 0

which looses the 192 ticks offset that the two substreams have relative to each other?

**Comment by sunghee-hwang** [comment]
I clarified it more on PR #186

> > end_timestamp[ID][N-1] = 0. For the first Sync OBU after join mid stream, N = 1.
> 
> Let's take two asynchronous substreams: ID=0 with 960 ticks per frame @ 48 kHz, and ID=1 with 1024 ticks per frame @ 48 kHz.
> 
> So the timeline for first 3 frames: ID=0 --> 0, 960, 1920 ID=1 --> 0, 1024, 2048
> 
> In absolute time after frame N=2 (after the third frames) then end_timestamp[ID=0][N=2] = 2880 end_timestamp[ID=1][N=2] = 3072.
> 
> If a playback system joins on frame N=3, then playback system will assume end_timestamp[ID=0][N=2] = 0 end_timestamp[ID=1][N=2] = 0
> 
> which looses the 192 ticks offset that the two substreams have relative to each other?

If you would put a Sync OBU between N = 2 and N = 3, the Sync OBU would provide the relative offsets which say -128 ticks for ID0 and 0 ticks for ID1 because max (end_timestamp[ID][N=2]) = 2048 , start_timestamp[ID=0][N=3] = 1920 and start_timestamp[ID=1][N=3] = 2048

**Comment by sunghee-hwang** [comment]
In this situation, if decoders join mid-stream starting from the above Sync OBU, then the Sync OBU is the first Sync OBU (i.e. N = 1) 
- end_timestamp[ID][0] = 0 for all ID.
- start_timestamp[ID=0][1] = -128
- start_timestamp[ID=1][1] = 0

**Comment by shawnisingh** [comment]
OK so I have come back to this and I am almost convinced about SungHee's proposed solution - thanks for the discussion a few weeks ago.   

Depending on our teminology, synchronizing timeline across a Sync OBU always requires some kind of procedure.  We can call that a concatenation rule.  So I think SungHee's proposal is really just proposing a new concatenation rule that may be able to replace our previous two rules.   That is separate from the choice of wording - do we still want to call it a concatenation rule in the spec or not.

Unfortunately I have a flood of other work today, so I'm requesting just another day to have a chance to think comprehensively about this.  I want to see if there is any value in our old concatenation rules which are not possible by the new concatenation rule, and I'd like to review the PR and make suggestions.   I hope it is OK to request just another day of time to have this discussion, thanks in advance!

**Comment by sunghee-hwang** [comment]
Resolved by PR #186 has been merged
